### PR TITLE
v1alpha4 types

### LINKED ISSUE #961: Create v1alpha4 API types
Tracking issue to generate v1alpha4 Cluster API types.

- change imports to use v1alpha4
- set storage version to v1alpha4 types
- remove old v1alpha2 types and conversion webhooks if necessary.

For reference: https://github.com/kubernetes-sigs/cluster-api/issues/3428

/kind api-change
/area provider/azure
/cc @nader-ziada

### PR DESCRIPTION
**Pull request by nader-ziada**
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:
- Generates types for v1alpha4
- Remove v1alpha2 types
- Updates imports to v1alpha4 packages

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #961 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] update min go version to 1.15.0
- [x] update controller-runtime to v0.8.1 when its merged in [capi](https://github.com/kubernetes-sigs/cluster-api/pull/4109) ?
- [x] update kind to v0.9.0
- [ ] what to do about v1alpha3 templates?

**Release note**:

```release-note
Add v1alpha4 types  
```

### CODE REVIEW

File: controllers/azurecluster_controller.go
Comment (nader-ziada) [inline controllers/azurecluster_controller.go:109]: not sure what if this is the right thing to do here with context since it now has to be in the signature of the func, in capi they set it in main.go

@devigned @CecileRobertMichon

File: controllers/azureidentity_controller.go
Comment (nader-ziada) [inline controllers/azureidentity_controller.go:85]: not sure what if this is the right thing to do here with context since it now has to be in the signature of the func, in capi they set it in main.go and in this case here was getting an error saying ctx is being updated on its first use

@devigned @CecileRobertMichon

File: controllers/azurecluster_controller.go
Comment (devigned) [reply controllers/azurecluster_controller.go:109]: We should also setup the signal handler in main.go. The signal handler listens to SIGTERM and SIGINT:
https://github.com/kubernetes-sigs/controller-runtime/blob/a763c9a36c6f18660799384c9765348942bda53a/pkg/manager/signals/signal.go#L30-L45

The intent of the loop timeout is to set a deterministic limit on the amount of time we will spend in a reconciliation loop. We should also keep the default loop timeout while passing in the outer context. 

```go
ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))`
```

File: controllers/azureidentity_controller.go
Comment (devigned) [reply controllers/azureidentity_controller.go:85]: Same as previous comment on context.

**Review by devigned — CHANGES_REQUESTED** [review]
Context guidance.

File: .golangci.yml
Comment (CecileRobertMichon) [inline .golangci.yml:43]: this is an exclude rule, shouldn't it exclude v1alpha3?

File: controllers/azurecluster_controller.go
Comment (nader-ziada) [reply controllers/azurecluster_controller.go:109]: fixed for all of them

File: controllers/azureidentity_controller.go
Comment (nader-ziada) [reply controllers/azureidentity_controller.go]: fixed

File: .golangci.yml
Comment (nader-ziada) [reply .golangci.yml]: yes, you are right, fixed it

File: docs/development.md
Comment (CecileRobertMichon) [inline docs/development.md:46]: 👍

File: exp/api/v1alpha4/groupversion_info.go
Comment (cpanato) [inline exp/api/v1alpha4/groupversion_info.go:2]: for new files, I think we should update the boilerplate date to be 2021

File: api/v1alpha3/azurecluster_types.go
Comment (cpanato) [inline api/v1alpha3/azurecluster_types.go:22]: why in the v1alpha3 we set to use the cluster-api v1alpha4 and not v1alpha3?

File: api/v1alpha3/azuremachine_conversion.go
Comment (cpanato) [inline api/v1alpha3/azuremachine_conversion.go:22]: this duplication is correct?

File: api/v1alpha3/azurecluster_types.go
Comment (nader-ziada) [reply api/v1alpha3/azurecluster_types.go:22]: you are correct, will fix it

File: api/v1alpha3/azuremachine_conversion.go
Comment (nader-ziada) [reply api/v1alpha3/azuremachine_conversion.go:22]: no, will fix it

File: azure/scope/cluster.go
Comment (CecileRobertMichon) [inline azure/scope/cluster.go:312]: it might be misleading that this is still returning a pointer but it's not a pointer to the spec value, it's a pointer to that subnet's route table which is itself not a pointer. Maybe it should just return a value, like ControlPlaneSubnet() does.

File: azure/scope/cluster_test.go
Comment (CecileRobertMichon) [inline azure/scope/cluster_test.go:34]: thank you for adding this

File: azure/scope/cluster_test.go
Comment (CecileRobertMichon) [inline azure/scope/cluster_test.go:70]: you shouldn't need to set env values here since we're using a fake authorizer

File: azure/scope/cluster_test.go
Comment (CecileRobertMichon) [inline azure/scope/cluster_test.go:88]: can we also test the case where custom rules have been set by the user?

**Review by CecileRobertMichon — COMMENTED** [review]
part 1 (still have 2/3rd of the files to go through)

File: test/e2e/azure_lb.go
Comment (CecileRobertMichon) [inline test/e2e/azure_lb.go:123]: why not pass in the ctx param here?

Comment (CecileRobertMichon) [reply test/e2e/azure_lb.go:123]: same question for the other test files in e2e

File: test/e2e/data/shared/v1alpha4/metadata.yaml
Comment (CecileRobertMichon) [inline test/e2e/data/shared/v1alpha4/metadata.yaml:6]: I thought there was discussion in CAPI office hours to keep clusterctl types at v1alpha3 for now

File: test/e2e/data/shared/v1alpha4_provider/metadata.yaml
Comment (CecileRobertMichon) [inline test/e2e/data/shared/v1alpha4_provider/metadata.yaml:6]: what is this file doing?

File: test/e2e/e2e_suite_test.go
Comment (CecileRobertMichon) [inline test/e2e/e2e_suite_test.go:161]: we should use ctx here

File: test/e2e/helpers.go
Comment (CecileRobertMichon) [inline test/e2e/helpers.go:63]: use ctx here and in the rest of this file

File: test/e2e/kubernetes/deployment/deployment.go
Comment (CecileRobertMichon) [inline test/e2e/kubernetes/deployment/deployment.go:118]: we should pass in a context here

File: test/e2e/kubernetes/namespace/namespace.go
Comment (CecileRobertMichon) [inline test/e2e/kubernetes/namespace/namespace.go:39]: same comment with context

**Review by CecileRobertMichon — COMMENTED** [review]
2/2 

Awesome work @nader-ziada

File: api/v1alpha3/types.go
Comment (nader-ziada) [reply api/v1alpha3/types.go:96]: yes, a few tests were failing which is great :D

File: api/v1alpha3/types.go
Comment (nader-ziada) [reply api/v1alpha3/types.go:160]: I think I covered this one as well

File: api/v1alpha4/azurecluster_validation.go
Comment (nader-ziada) [reply api/v1alpha4/azurecluster_validation.go:101]: I will add a field error there if the err != nil

File: api/v1alpha4/tags.go
Comment (nader-ziada) [reply api/v1alpha4/tags.go:138]: this is the same as in the main branch v1alpha3 file: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/tags.go#L134-L138 

did that get messed up somehow?

File: azure/scope/cluster.go
Comment (nader-ziada) [reply azure/scope/cluster.go:486]: sorry about all those :D

File: controllers/azurecluster_controller_test.go
Comment (nader-ziada) [reply controllers/azurecluster_controller_test.go:72]: I think `gomock.AContext()` is a matcher used to validate that a context is passed in a mock func, but in this case here we are calling an actual func. I think test are supposed to call context.TODO() so I can change to that

File: scripts/ci-e2e.sh
Comment (nader-ziada) [reply scripts/ci-e2e.sh:77]: yes

File: scripts/kind-with-registry.sh
Comment (nader-ziada) [reply scripts/kind-with-registry.sh:26]: yes

File: test/e2e/data/shared/v1alpha4_provider/metadata.yaml
Comment (nader-ziada) [reply test/e2e/data/shared/v1alpha4_provider/metadata.yaml:6]: it will be used by clusterctl as far as I understand

File: test/e2e/azure_lb.go
Comment (nader-ziada) [reply test/e2e/azure_lb.go:123]: yes, it should use ctx

File: test/e2e/kubernetes/namespace/namespace.go
Comment (nader-ziada) [reply test/e2e/kubernetes/namespace/namespace.go:39]: will do

File: test/e2e/kubernetes/deployment/deployment.go
Comment (nader-ziada) [reply test/e2e/kubernetes/deployment/deployment.go:118]: will do

File: test/e2e/helpers.go
Comment (nader-ziada) [reply test/e2e/helpers.go:63]: yup

File: test/e2e/e2e_suite_test.go
Comment (nader-ziada) [reply test/e2e/e2e_suite_test.go:161]: will do

File: controllers/azurejson_machine_controller.go
Comment (nader-ziada) [reply controllers/azurejson_machine_controller.go:87]: should be removed

File: api/v1alpha3/azurecluster_conversion.go
Comment (nader-ziada) [reply api/v1alpha3/azurecluster_conversion.go:40]: The subnet can be automatically converted, the test fails 
```
--- FAIL: TestFuzzyConversion/for_AzureCluster (0.00s)
        --- FAIL: TestFuzzyConversion/for_AzureCluster/hub-spoke-hub (0.00s)
            conversion.go:193: 
                  &v1alpha4.AzureCluster{
                        TypeMeta:   {},
                        ObjectMeta: {Name: "ȴ,]始侏Md}.", GenerateName: `dQ"8ʗ~ǰĵïF`, Namespace: "{喾除e湽ǛŮɐŪ毄謖橊E;槊賎硻轲I", SelfLink: "晻軾Ĳ蜂塒¦ȋË稻", ...},
                        Spec: v1alpha4.AzureClusterSpec{
                                NetworkSpec: v1alpha4.NetworkSpec{
                                        Vnet:    {ResourceGroup: "U瘟ȖɟǴ业獳oFB>ƽ", ID: "ŀ赐ȅƤɲ铩X槎Ö", Name: "龀蝡Aƃƛ沶掁Ȫ", CidrBlock: "ƈ淄", ...},
                -                       Subnets: nil,
                +                       Subnets: v1alpha4.Subnets{},
                                        APIServerLB: v1alpha4.LoadBalancerSpec{
                -                               ID:          "ɼ犄红",
                +                               ID:          "",
                -                               Name:        "k`ǈ斞riȡƦƈ衝ůlģ",
                +                               Name:        "",
                -                               SKU:         `ɮ"£<5`,
                +                               SKU:         "",
                                                FrontendIPs: nil,
                -                               Type:        "培Ų@ǿ輁Ž0-瓋莹ĘĴ>",
                +                               Type:        "",
                                        },
                                },
                                ResourceGroup:  ` 裌泖m1庸;B|ŗ琭\ŝɁê d`,
                                SubscriptionID: "Ƿȍʜ晊遟i羏崜.离JɌ",
                                ... // 4 identical fields
                        },
                        Status: {Ready: true},
                  }
                
                Expected
                    <bool>: false
                to be true
```

File: api/v1alpha3/azuremachine_conversion.go
Comment (nader-ziada) [reply api/v1alpha3/azuremachine_conversion.go:106]: I don't remember why I added this func, but removed it and everything seems fine

File: azure/scope/cluster.go
Comment (nader-ziada) [reply azure/scope/cluster.go:312]: yeah, what you are saying makes sense, will make the func return the spec type without a pointer

File: azure/scope/cluster_test.go
Comment (nader-ziada) [reply azure/scope/cluster_test.go]: will remove them

**Comment by nader-ziada** [comment]
@CecileRobertMichon addressed most of your comments in a separate commit (for now) in case you want to check them

**Comment by nader-ziada** [comment]
the `tilt up` is currently not working for some reason on this branch, need to investigate further

File: test/e2e/kubernetes/networkpolicy/networkpolicy.go
Comment (devigned) [inline test/e2e/kubernetes/networkpolicy/networkpolicy.go:64]: Please pipe through a context, or create an issue to follow up on piping through contexts for each place a `context.Background()` was added. Each one is an opportunity for the func to block indefinitely (or until the clientset timeout occurs).

**Review by devigned — CHANGES_REQUESTED** [review]
Wow... That was an absurd amount of changes for an API version. It is not the fault of the PR, but rather all the code generation and import path manipulations that are required.

Great work with this PR! This was obviously a significant effort.

I've left mostly comments, but a few requests for change. Particularly, I found the use of `context.Background` in the e2e test framework to be a problem. I don't want to make this PR larger than it already is, so if we want to file a follow up issue, I'd support it.

Thank you, @nader-ziada!

File: api/v1alpha4/azuremachine_types.go
Comment (CecileRobertMichon) [reply api/v1alpha4/azuremachine_types.go:46]: let's do that separately, there are a number of changes needed in v1alpha4 types but this PR is just doing a lift and shift

Comment (nader-ziada) [reply api/v1alpha4/azuremachine_types.go:46]: +1 to a follow up PR for all the `deprecated`

File: test/e2e/kubernetes/networkpolicy/networkpolicy.go
Comment (nader-ziada) [reply test/e2e/kubernetes/networkpolicy/networkpolicy.go]: yes will do that

**Comment by CecileRobertMichon** [comment]
lgtm overall

Let's get this pulled in and iterate

is tilt working now?

**Comment by nader-ziada** [comment]
> lgtm overall
> 
> Let's get this pulled in and iterate
> 
> is tilt working now?

not yet, can't figure out what'w wrong with it, was just looking at it

**Comment by nader-ziada** [comment]
@CecileRobertMichon I think tilt is fine now

**Comment by nader-ziada** [comment]
@CecileRobertMichon squashed commits

**Comment by CecileRobertMichon** [comment]
/retest

@nader-ziada are we good to remove the WIP from PR title?

File: Tiltfile
Comment (CecileRobertMichon) [inline Tiltfile:160]: why remove these?

Comment (nader-ziada) [reply Tiltfile:160]: tilt was complaining about this line saying I doesn't have a a deployment, don't know if I need to define it somewhere else

**Comment by CecileRobertMichon** [comment]
@devigned have all your comments been addressed or filed as follow-up issues?

If so let's merge this after #1180 merges. 

/lgtm
/hold

File: test/e2e/kubernetes/node/node.go
Comment (devigned) [inline test/e2e/kubernetes/node/node.go:35]: Looks like one last context. Either a follow up issue or one more context plumbing job.

**Review by devigned — CHANGES_REQUESTED** [review]
Found one more context nit which we can either follow up on or fix here. Outside of that, this is good to go from my point of view.

File: test/e2e/azure_lb.go
Comment (devigned) [inline test/e2e/azure_lb.go:202]: Some more contexts :(

**Review by devigned — CHANGES_REQUESTED** [review]

File: test/e2e/azure_lb.go
Comment (nader-ziada) [reply test/e2e/azure_lb.go]: fixed

**Review by devigned — APPROVED** [review]
/lgtm

Awesome work! :rocket:

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
